### PR TITLE
feat(auth): Hosted UI

### PR DESCRIPTION
**Pull request by dnys1**
- Dart-only Hosted UI platform and state machine
- iOS/Android bridge for platform webviews

commit-id:7c629b62

---

**Stack**:
- #1677
- #1676
- #1675
- #1670
- #1655
- #1654
- #1652
- #1648 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: packages/auth/amplify_auth_cognito/pigeons/native_auth_plugin.dart [hunk 1]
```diff
@@ -45,8 +45,33 @@ abstract class NativeAuthPlugin {
...
 @HostApi()
 abstract class NativeAuthBridge {
+  /// Configure the native platform/plugin.
+  ///
+  /// On iOS/Android, this calls `Amplify.addPlugin` with the [NativeAuthPlugin]
+  /// implementation.
   @async
   void configure();
+
+  /// Sign out by presenting [url] and waiting for a response to a URL with
```
Comment (Jordan-Nelson) [inline packages/auth/amplify_auth_cognito/pigeons/native_auth_plugin.dart:55]:
```suggestion
  /// Sign in by presenting [url] and waiting for a response to a URL with
```

File: packages/auth/amplify_auth_cognito/pigeons/native_auth_plugin.dart [hunk 2]
```diff
@@ -45,8 +45,33 @@ abstract class NativeAuthPlugin {
...
+  /// [callbackUrlScheme].
+  ///
+  /// If [preferPrivateSession] is `true`, do not persist session cookies.
+  @async
+  Map<String, String> signIn(
```
Comment (Jordan-Nelson) [inline packages/auth/amplify_auth_cognito/pigeons/native_auth_plugin.dart:60]: *nit:* `signInWithUrl` would make it clear that this has nothing to do with username/password auth.

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AtomicResult.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import io.flutter.Log
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.launch
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * Thread-safe [NativeAuthPluginBindings.Result] wrapper which prevents multiple replies and
+ * automatically posts results to the main thread.
+ */
+class AtomicResult<T>(
```
Comment (Jordan-Nelson) [inline packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AtomicResult.kt:26]: Should the one in core eventually be replaced with this one?

**Review by Jordan-Nelson — APPROVED** [review]

File: packages/amplify/amplify_flutter/lib/src/hybrid_impl.dart [hunk 4]
```diff
@@ -80,6 +80,8 @@ class AmplifyHybridImpl extends AmplifyClass {
               AmplifyExceptionMessages.missingRecoverySuggestion,
         );
       }
+    } on AmplifyAlreadyConfiguredException {
+      return;
```
Comment (haverchuck) [inline packages/amplify/amplify_flutter/lib/src/hybrid_impl.dart:84]: Maybe we want to set the isConfigured flag to true here?

Comment (haverchuck) [inline packages/amplify/amplify_flutter/lib/src/hybrid_impl.dart:84]: Maybe we want to set the isConfigured flag to true here?

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AtomicResult.kt (hunk 3 above)
Comment (dnys1) [reply packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AtomicResult.kt:26]: This wraps the generated bindings' result which is different than a generic Flutter result. So they're incompatible and we'll need both, unfortunately, unless at some point we migrate all the plugins to a single pigeon definition in core which could be pretty cool.

File: packages/auth/amplify_auth_cognito/lib/src/flows/hosted_ui/hosted_ui_platform_html.dart [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  /// The base URL
+  String get baseUrl {
+    final basePath = getBaseElementHrefFromDom() ?? '/';
+    return url.join(window.location.origin, basePath);
+  }
+
+  @override
+  Uri get signInRedirectUri => config.signInRedirectUris.firstWhere(
+        (uri) => uri.toString().startsWith(baseUrl),
+        orElse: () => throw UnknownException(
```
Comment (haverchuck) [inline packages/auth/amplify_auth_cognito/lib/src/flows/hosted_ui/hosted_ui_platform_html.dart:37]: Do we want to have some sort `InvalidConfiguration` exception?

File: packages/auth/amplify_auth_cognito/pubspec.yaml [hunk 6]
```diff
@@ -12,6 +12,7 @@ dependencies:
   amplify_auth_cognito_ios: 0.5.0
   amplify_auth_plugin_interface: 0.5.0
   amplify_core: 0.5.0
+  amplify_flutter: 0.5.0
```
Comment (haverchuck) [inline packages/auth/amplify_auth_cognito/pubspec.yaml:15]: Can we avoid this? I think it's preferable to avoid circular dependencies.

**Review by haverchuck — COMMENTED** [review]
Left some questions/comments, but mostly LGTM

File: packages/auth/amplify_auth_cognito/pubspec.yaml (hunk 6 above)
Comment (dnys1) [reply packages/auth/amplify_auth_cognito/pubspec.yaml:15]: amplify_flutter doesn't depend on amplify_auth_cognito, not sure what circular dependency you mean

File: packages/amplify/amplify_flutter/lib/src/hybrid_impl.dart (hunk 4 above)
Comment (dnys1) [reply packages/amplify/amplify_flutter/lib/src/hybrid_impl.dart:84]: It's set by completing asyncConfig. If we hot restart these will throw but by ignoring them, asyncConfig will be set

File: packages/auth/amplify_auth_cognito/pubspec.yaml (hunk 6 above)
Comment (haverchuck) [reply packages/auth/amplify_auth_cognito/pubspec.yaml:15]: I think I may have been thinking about the old interface... this is fine.

**Review by haverchuck — APPROVED** [review]

File: packages/auth/amplify_auth_cognito/lib/src/flows/hosted_ui/hosted_ui_platform_html.dart (hunk 5 above)
Comment (dnys1) [reply packages/auth/amplify_auth_cognito/lib/src/flows/hosted_ui/hosted_ui_platform_html.dart:37]: We have InvalidUserPoolConfiguration which I guess works. I'll update it.

**Pull request merged**